diff --git a/src/lib/process/env.c b/src/lib/process/env.c
index 0060200..3184a05 100644
--- a/src/lib/process/env.c
+++ b/src/lib/process/env.c
@@ -28,28 +28,28 @@
 #include <crt_externs.h>
 #endif
 
-#ifndef HAVE__NSGETENVIRON
+//#ifndef HAVE__NSGETENVIRON
 #ifndef HAVE_EXTERN_ENVIRON_DECLARED
 /* Some platforms declare environ under some circumstances, others don't. */
 #ifndef RUNNING_DOXYGEN
 extern char **environ;
 #endif
 #endif /* !defined(HAVE_EXTERN_ENVIRON_DECLARED) */
-#endif /* !defined(HAVE__NSGETENVIRON) */
+//#endif /* !defined(HAVE__NSGETENVIRON) */
 
 /** Return the current environment. This is a portable replacement for
  * 'environ'. */
 char **
 get_environment(void)
 {
-#ifdef HAVE__NSGETENVIRON
+//#ifdef HAVE__NSGETENVIRON
   /* This is for compatibility between OSX versions.  Otherwise (for example)
    * when we do a mostly-static build on OSX 10.7, the resulting binary won't
    * work on OSX 10.6. */
-  return *_NSGetEnviron();
-#else /* !(defined(HAVE__NSGETENVIRON)) */
+//  return *_NSGetEnviron();
+//#else /* !(defined(HAVE__NSGETENVIRON)) */
   return environ;
-#endif /* defined(HAVE__NSGETENVIRON) */
+//#endif /* defined(HAVE__NSGETENVIRON) */
 }
 
 /** Helper: return the number of characters in <b>s</b> preceding the first
